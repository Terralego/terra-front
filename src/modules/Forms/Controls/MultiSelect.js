import React from 'react';

import PropTypes from 'prop-types';

import classnames from 'classnames';
import {
  Button,
  Intent,
  Menu,
  MenuItem,
  Spinner,
} from '@blueprintjs/core';
import { MultiSelect as BPMultiSelect } from '@blueprintjs/select';

import { preventEnterKeyPress } from '../../../utils/event';
import translateMock from '../../../utils/translate';
import NoValues from './NoValues';
import formatValues from './formatValues';

// With more than 250 items we don't display result but we propose a search
const MAX_DISPLAY_ITEMS = 250;

export class MultiSelect extends React.Component {
  static propTypes = {
    value: PropTypes.arrayOf(PropTypes.string),
    label: PropTypes.string,
    onChange: PropTypes.func,
    values: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          label: PropTypes.string.isRequired,
          value: PropTypes.any.isRequired,
        }),
      ]),
    ).isRequired,
    isSubmissionPrevented: PropTypes.bool,
    translate: PropTypes.func,
    loading: PropTypes.bool,
    minCharacters: PropTypes.number,
    highlightSearch: PropTypes.bool,
  };

  static defaultProps = {
    label: '',
    value: [],
    isSubmissionPrevented: true,
    onChange () {},
    translate: translateMock({
      'terralego.forms.controls.multiselect.no_results': 'No results.',
      'terralego.forms.controls.multiselect.select_placeholder': 'Enter a query first',
      'terralego.forms.controls.multiselect.select_placeholder_plural': 'Enter at least {{count}} characters',
      'terralego.forms.controls.multiselect.input_placeholder': 'Search...',
      'terralego.forms.controls.multiselect.too_many': 'Too many results, please refine your query...',
      'terralego.forms.controls.noValues': 'No choice available',
    }),
    loading: false,
    minCharacters: 0,
    highlightSearch: true,
  };

  state = {
    values: [],
    value: [],
  }

  static getDerivedStateFromProps ({ values, value }) {
    const formatedValues = formatValues(values);

    const formatedValue = value.map(val => ({
      label: formatedValues.find(({ value: availableVal }) => val === availableVal).label,
      value: val,
    }));

    return {
      values: formatedValues,
      value: formatedValue,
    };
  }

  handleChange = item => {
    const { value, onChange } = this.props;
    const newValue = value.includes(item.value)
      ? [...value.filter(val => val !== item.value)]
      : [...value, item.value];
    onChange(newValue);
  };

  handleTagRemove = (_, index) => {
    const { value, onChange } = this.props;
    const newValue = value.filter((__, idx) => idx !== index);
    onChange(newValue);
  }

  handleClear = () => {
    const { onChange } = this.props;
    onChange([]);
  };

  render () {
    const {
      handleChange,
      handleTagRemove,
    } = this;

    const {
      label,
      value: origValue,
      isSubmissionPrevented,
      className,
      translate,
      loading,
      minCharacters,
      highlightSearch,
      ...props
    } = this.props;

    const { values, value } = this.state;
    const displayClearButton = value.length > 0;

    return (
      <div
        className="control-container"
        onKeyPress={isSubmissionPrevented ? preventEnterKeyPress : null}
        role="presentation"
      >
        <p className="control-label">{label}</p>
        {loading && <Spinner size={20} />}
        {!loading && !values.length && <NoValues placeholder={translate('terralego.forms.controls.noValues')} />}
        {!loading && (
          <BPMultiSelect
            className={classnames('tf-multiselect', className)}
            resetOnSelect
            fill
            items={values}
            selectedItems={value}
            onItemSelect={handleChange}
            tagRenderer={item => item.label}
            popoverProps={{ usePortal: false }}
            placeholder={translate('terralego.forms.controls.multiselect.input_placeholder')}
            tagInputProps={{
              tagProps: { intent: Intent.NONE, interactive: true },
              onRemove: handleTagRemove,
              rightElement: displayClearButton && (
                <Button icon="cross" minimal onClick={this.handleClear} />
              ),
            }}
            itemListPredicate={(query, items) => {
              const regExp = new RegExp(query, 'i');
              return query.length < minCharacters
                ? items
                : items.filter(item => regExp.test(item.label));
            }}
            itemRenderer={({ label: itemLabel, value: itemValue }, {
              query, handleClick, modifiers: { matchesPredicate, ...modifiers },
            }) => {
              if (!matchesPredicate) {
                return null;
              }
              const regExp = new RegExp(`(${query})`, 'i');
              return (
                <MenuItem
                  active={modifiers.active}
                  icon={origValue.includes(itemValue) ? 'tick' : 'blank'}
                  key={`${itemValue}${itemLabel}`}
                  onClick={handleClick}
                  text={query && highlightSearch
                    ? itemLabel.split(regExp).map(
                      // eslint-disable-next-line react/no-array-index-key
                      (elem, i) => (i % 2 ? <strong key={`${itemLabel}-${i}`}>{elem}</strong> : elem),
                    ) : itemLabel}
                  shouldDismissPopover={false}
                />
              );
            }}
            itemListRenderer={listProps => {
              const items = listProps.filteredItems;
              let message = null;
              // If we have no query or a query too short and too many results
              if ((!listProps.query || listProps.query.length < minCharacters)
                && items.length > MAX_DISPLAY_ITEMS) {
                message = translate('terralego.forms.controls.multiselect.select_placeholder', { count: minCharacters });
              } else if (items.length > MAX_DISPLAY_ITEMS) {
                // If we have a query but too many results
                message = translate('terralego.forms.controls.multiselect.too_many');
              } else if (items.length === 0) {
                // If we have no results
                message = translate('terralego.forms.controls.multiselect.no_results');
              }
              return (
                <Menu ulRef={listProps.itemsParentRef}>
                  {message !== null
                    ? (
                      <MenuItem
                        disabled
                        text={message}
                      />
                    )
                    : items.map(listProps.renderItem)
                  }
                </Menu>
              );
            }}
            {...props}
          />
        )}
      </div>
    );
  }
}

export default MultiSelect;
