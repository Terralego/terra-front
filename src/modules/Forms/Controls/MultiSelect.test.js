import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import MultiSelect from './MultiSelect';

jest.mock('@blueprintjs/select', () => ({
  MultiSelect: function BPMultiSelect ({ items }) {
    return (
      <select multiple>
        {items.map(({ label, value }) => (
          <option key={label} value={value}>{label}</option>
        ))}
      </select>
    );
  },
}));

jest.mock('@blueprintjs/core', () => ({
  Button ({ children }) {
    return <div className="Button"> {children} </div>;
  },
  Menu: function BPMenu ({ children }) {
    return <div className="Menu"> {children} </div>;
  },
  MenuItem: function BPMenuItem ({ children }) {
    return <div className="MenuItem"> {children} </div>;
  },
  FormGroup ({ children }) {
    return <div className="FormGroup"> {children} </div>;
  },
  Intent: {
    NONE: 'none',
  },
  Spinner: () => <p>Spinner</p>,
}));

it('should render correctly', () => {
  const tree = renderer.create((
    <MultiSelect
      label="Bouh"
      values={['foo', 'bar']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render loading correctly', () => {
  const tree = renderer.create((
    <MultiSelect label="Pwout" loading values={[]} />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render with values', () => {
  const tree = renderer.create((
    <MultiSelect
      label="Bouh"
      values={['foo', 'bar']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render multiple value', () => {
  const tree = renderer.create((
    <MultiSelect
      label="Bouh"
      values={['foo', 'bar']}
      value={['foo', 'bar']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();

  const tree2 = renderer.create((
    <MultiSelect
      label="Boom"
      values={[{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }]}
      value={['foo', 'bar']}
    />
  ));
  expect(tree2.toJSON()).toMatchSnapshot();
});

it('should handle change', () => {
  const onChange = jest.fn();
  const instance = new MultiSelect({ onChange, value: [], values: ['foo'] });
  instance.handleChange({ value: 'foo' });
  expect(onChange).toHaveBeenCalledWith(['foo']);
  onChange.mockClear();

  instance.props.value = ['foo'];
  instance.handleChange({ value: instance.props.value[0] });
  expect(onChange).toHaveBeenCalledWith([]);
});

it('should handle clear', () => {
  const onChange = jest.fn();
  const instance = new MultiSelect({ onChange });
  instance.handleClear();
  expect(onChange).toHaveBeenCalledWith([]);
});

it('should handle tag remove', () => {
  const onChange = jest.fn();
  const instance = new MultiSelect({ onChange, value: [] });
  instance.props.value = ['foo', 'bar'];
  instance.handleTagRemove('foo', 0);

  expect(onChange).toHaveBeenCalledWith(['bar']);
});

it('should render tag', () => {
  const wrapper = shallow(<MultiSelect values={['foo', 'bar']} />);
  const { tagRenderer } = wrapper.find('BPMultiSelect').props();
  expect(tagRenderer({ value: 'foo', label: 'Foo' })).toBe('Foo');
});

it('should render item', () => {
  const wrapper = shallow(<MultiSelect values={[{ label: 'Foo', value: 'foo' }, { label: 'Bar', value: 'bar' }]} />);
  {
    const { itemRenderer } = wrapper.find('BPMultiSelect').props();
    expect(itemRenderer('foo', { modifiers: { matchesPredicate: false } })).toBe(null);
    expect(itemRenderer('foo', { modifiers: { matchesPredicate: true } })).not.toBe(null);
    const itemRendered = itemRenderer({ label: 'Foo', value: 'foo' }, { modifiers: { matchesPredicate: true } });
    expect(itemRendered.type.name).toBe('BPMenuItem');
    expect(itemRendered.props.icon).toBe('blank');
  }
  wrapper.setProps({ value: ['foo'] });
  {
    const { itemRenderer } = wrapper.find('BPMultiSelect').props();
    const itemRendered = itemRenderer({ label: 'Foo', value: 'foo' }, { modifiers: { matchesPredicate: true } });
    expect(itemRendered.props.icon).toBe('tick');
  }
});

it('should render item with highlighted support', () => {
  const wrapper = shallow(<MultiSelect values={[{ label: 'Foo', value: 'foobar' }, { label: 'Bar', value: 'barbaz' }]} />);
  {
    const { itemRenderer } = wrapper.find('BPMultiSelect').props();
    const itemRendered = itemRenderer({ label: 'foobarbaz', value: 'foobarbaz' }, { query: 'bar', modifiers: { matchesPredicate: true } });
    expect(itemRendered.props.text).toEqual(['foo', <strong key="foobarbaz-1">bar</strong>, 'baz']);
  }
  wrapper.setProps({ highlightSearch: false });
  {
    const { itemRenderer } = wrapper.find('BPMultiSelect').props();
    const itemRendered = itemRenderer({ label: 'foobar', value: 'foobar' }, { query: 'foo', modifiers: { matchesPredicate: true } });
    expect(itemRendered.props.text).toBe('foobar');
  }
});

it('should render correctly default message', () => {
  const renderItem = jest.fn();
  const wrapper = shallow(<MultiSelect values={['foo', 'bar']} />);
  expect(shallow(wrapper.find('BPMultiSelect').props().itemListRenderer({
    filteredItems: [],
    query: 'something',
  })).find('BPMenuItem').props().text).toBe('No results.');
  expect(shallow(wrapper.find('BPMultiSelect').props().itemListRenderer({
    filteredItems: Array(251).fill('foo'),
    query: 'fo',
  })).find('BPMenuItem').props().text).toBe('Too many results, please refine your query...');
  expect(shallow(wrapper.find('BPMultiSelect').props().itemListRenderer({
    filteredItems: Array(251).fill('foo'),
    query: '',
  })).find('BPMenuItem').props().text).toBe('Enter a query first');
  wrapper.find('BPMultiSelect').props().itemListRenderer({
    filteredItems: ['foo', 'bar'],
    query: '',
    renderItem,
  });
  expect(renderItem).toHaveBeenCalledTimes(2);
});

it('should prevent submit event in parent', () => {
  const wrapper = shallow(<MultiSelect values={['foo', 'bar']} isSubmissionPrevented={false} />);
  const div = wrapper.find('.control-container');
  expect(div.props().onKeyPress).toBe(null);
});

it('should not filter if minCharacters not matched', () => {
  const values = ['foo', 'bar'];
  const wrapper = shallow(<MultiSelect values={values} minCharacters={2} />);
  expect(wrapper.find('BPMultiSelect').props().itemListPredicate('f', values)).toBe(values);
  expect(wrapper.find('BPMultiSelect').props().itemListPredicate('fo', values)).toEqual(['foo']);
});
