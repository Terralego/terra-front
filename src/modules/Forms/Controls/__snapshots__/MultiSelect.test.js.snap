// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Bouh
  </p>
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
</div>
`;

exports[`should render loading correctly 1`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <p>
    Spinner
  </p>
</div>
`;

exports[`should render multiple value 1`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Bouh
  </p>
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
</div>
`;

exports[`should render multiple value 2`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Boom
  </p>
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      Foo
    </option>
    <option
      value="bar"
    >
      Bar
    </option>
  </select>
</div>
`;

exports[`should render with values 1`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Bouh
  </p>
  <select
    multiple={true}
  >
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="bar"
    >
      bar
    </option>
  </select>
</div>
`;
